src/toil/lib/toilDocker.py

```diff
@@ -0,0 +1,213 @@
+           If list of lists: list[list[str]], then treat as successive commands chained with pipe.
+    :param str workDir: Directory to mount into the container via `-v`. Destination convention is /data
+    :param list[str] dockerParameters: Parameters to pass to Docker. Default parameters are `--rm`,
+            `--log-driver none`, and the mountpoint `-v work_dir:/data` where /data is the destination convention.
+             These defaults are removed if docker_parmaters is passed, so be sure to pass them if they are desired.
+    :param file outfile: Pipe output of Docker call to file handle
+    :param int deferParam: What action should be taken on the container upon job completion?
+           FORGO (0) will leave the container untouched.
+           STOP (1) will attempt to stop the container with `docker stop` (useful for debugging).
+           RM (2) will stop the container and then forcefully remove it from the system
+           using `docker rm -f`. This is the default behavior if defer is set to None.
+    """
+    thisUser = os.getuid()
+    thisGroup = os.getgid()
+
+    if containerName is None:
+        containerName = _getContainerName(job)
+    if workDir is None:
+        workDir = os.getcwd()
+
+    # If 'parameters' is a list of lists, treat each list as a separate command and
+    # chain with pipes.
+    if len(parameters) > 0 and type(parameters[0]) is list:
+        chain_params = [' '.join(p) for p in [map(pipes.quote, q) for q in parameters]]
+        command = ' | '.join(chain_params)
+        pipe_prefix = "set -eo pipefail && "
+        command = [pipe_prefix + command]
+        _logger.info("Calling docker with: " + repr(command))
+
+    # If 'parameters' is a normal list, join all elements into a single string element
+    # Example: ['echo','the', 'Oread'] becomes: ['echo the Oread']
+    # Note that this is still a list, and the docker API prefers this as best practice:
+    # http://docker-py.readthedocs.io/en/stable/containers.html
+    elif len(parameters) > 0 and type(parameters) is list:
+        command = [' '.join(p) for p in [map(pipes.quote, parameters)]]
+        _logger.info("Calling docker with: " + repr(command))
+
+    # If the 'parameters' lists are empty, they are respecified as None, which tells the API
+    # to simply create and run the container
+    else:
+        entrypoint = None
+        command = None
+
+    workDir = os.path.abspath(workDir)
+
+    # Ensure the user has passed a valid value for deferParam
+    assert deferParam in (None, FORGO, STOP, RM), 'Please provide a valid value for deferParam.'
+
+    client = docker.from_env()
+
+    if deferParam == STOP:
+        job.defer(_dockerStop, containerName, client)
+
+    if deferParam == FORGO:
+        removeUponExit = False
+    elif deferParam == RM:
+        removeUponExit = True
+        job.defer(_dockerKill, containerName, client)
+    # really shaky on this ->>
```

Comment:
On what?

Reply:
It overrides the docker API behavior in such a way that it's impossible to simply call docker with the "--rm" option only.  I'll try to think of a better way implement this.